core/os/thread.cpp

```diff
@@ -50,21 +50,27 @@ void Thread::_set_platform_functions(const PlatformFunctions &p_functions) {
 	platform_functions = p_functions;
 }
 
-void Thread::callback(Thread *p_self, const Settings &p_settings, Callback p_callback, void *p_userdata) {
-	Thread::caller_id = _thread_id_hash(p_self->thread.get_id());
+void Thread::callback(ID p_caller_id, const Settings &p_settings, Callback p_callback, void *p_userdata) {
+	Thread::caller_id = p_caller_id;
 	if (platform_functions.set_priority) {
 		platform_functions.set_priority(p_settings.priority);
 	}
 	if (platform_functions.init) {
 		platform_functions.init();
 	}
-	ScriptServer::thread_enter(); // Scripts may need to attach a stack.
+
+	// Currently causes a race condition due to the fact that threads are
+	// spawned before ScriptServer is even initialized.
+	// This is currently unused since neither GDScript nor C# require
+	// attaching a stack to new threads, so we just disable for now.
+
+	//ScriptServer::thread_enter();
```

Comment:
This might break potential ScriptLanguage bindings via GDExtension?

The other changes seem to be clear fixes and could be cherry-picked to `4.0`, but this looks more like a temporary hack so I'd suggest moving it to a new PR.

Reply:
@akien-mga 
See the comments above for the specifics,[^link] but in summary:
[^link]: [#issuecomment-1442637066](#issuecomment-1442637066)
  [#issuecomment-1442891021](#issuecomment-1442891021)
  [#issuecomment-1443829611](#issuecomment-1443829611)

- `WorkerThreadPool` is already broken for those languages
- The only available workarounds/fixes are much more involved (such as rearranging `main()`), and would probably break something else
- Having ThreadSanitizer in CI first would help catch any problems with said workaround/fix

Given that all options involve breakage, I would opt for the one that at least [enables ThreadSanitizer to be added into CI sooner](73777), since this is currently preventing that.

What are your thoughts? Does this warrant further discussion in a new PR?

Reply:
My idea would be that the other changes can be merged ASAP and cherry-picked for 4.0.1, as they seem to be clear improvements/fixes.

This part about ScriptLanguage requires IMO more discussion / to be handled separately. I'm also fine with merging it ASAP in `master` so that we can enable ThreadSanitizer on CI (I agree that's a worthy goal), but I wouldn't cherry-pick it to `4.0` until there's a clear consensus on it being a correct change (which would ideally go all the way and remove `ScriptServer::thread_enter`, etc. if not used). So splitting that to a separate PR allows me to cherry-pick the other changes while leaving the ScriptLanguage part as is in the short term in `4.0`.